### PR TITLE
[data][fault tolerance] Do not eagerly free root RefBundles

### PR DESCRIPTION
**Pull request by raulchen**
## Why are these changes needed?

Fix the issue that we eagerly free the root RefBundles. This bug broke fault tolerance.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/planner/plan_read_op.py [hunk 1]
```diff
@@ -93,25 +93,24 @@ def plan_read_op(op: Read) -> PhysicalOperator:
...
+        RefBundle(
+            [
+                (
+                    # TODO(chengsu): figure out a better way to pass read
+                    # tasks other than ray.put().
+                    ray.put(read_task),
+                    cleaned_metadata(read_task),
+                )
+            ],
+            owns_blocks=True,
```
Comment (ericl) [inline python/ray/data/_internal/planner/plan_read_op.py:107]: If you set owns_blocks=False that would also equivalently prevent them from being freed.

Comment (ericl) [reply python/ray/data/_internal/planner/plan_read_op.py:107]: Though, it sounds like we'd like to free them at the end of the workload, just not as soon as possible, so there is some subtlety here.

Comment (raulchen) [reply python/ray/data/_internal/planner/plan_read_op.py:107]: yes, `owns_blocks=False` is the right fix. at the end of the workload, they will be eventually freed by python GC. and their sizes are small. So it should be fine.

**Review by ericl — APPROVED** [review]

File: python/ray/data/_internal/planner/plan_read_op.py [hunk 2]
```diff
@@ -105,7 +105,10 @@ def get_input_data() -> List[RefBundle]:
                         cleaned_metadata(read_task),
                     )
                 ],
-                owns_blocks=True,
+                # `owns_blocks` is False, because these refs are the root of the DAG.
+                # We shouldn't eagerly free them. Otherwise, the DAG cannot be
+                # reconstructed.
+                owns_blocks=False,
```
Comment (c21) [inline python/ray/data/_internal/planner/plan_read_op.py:111]: should we also change of legacy compat code path - https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/execution/legacy_compat.py#L223 ?

**Review by c21 — APPROVED** [review]

**Review by amogkam — APPROVED** [review]

**Pull request merged**